File: st-exerciseviewer/src/main/java/de/saring/exerciseviewer/gui/panels/SamplePanelController.java
package de.saring.exerciseviewer.gui.panels;

import javafx.collections.FXCollections;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.control.cell.PropertyValueFactory;

import de.saring.exerciseviewer.data.ExerciseSample;
import de.saring.exerciseviewer.data.RecordingMode;
import de.saring.exerciseviewer.gui.EVContext;
import de.saring.util.gui.javafx.FormattedNumberCellFactory;

import javax.inject.Inject;

/**
 * Controller (MVC) class of the "Samples" panel, which displays all recorded samples of the exercise in a table.
 *
 * @author Stefan Saring
 */
public class SamplePanelController extends AbstractPanelController {

    @FXML
    private TableView<ExerciseSample> tvSamples;

    @FXML
    private TableColumn<ExerciseSample, Number> tcTime;
    @FXML
    private TableColumn<ExerciseSample, Number> tcHeartrate;
    @FXML
    private TableColumn<ExerciseSample, Number> tcAltitude;
    @FXML
    private TableColumn<ExerciseSample, Number> tcSpeed;
    @FXML
    private TableColumn<ExerciseSample, Number> tcDistance;
    @FXML
    private TableColumn<ExerciseSample, Number> tcCadence;
    @FXML
    private TableColumn<ExerciseSample, Number> tcTemperature;

    /**
     * Standard c'tor for dependency injection.
     *
     * @param context the ExerciseViewer UI context
     */
    @Inject
    public SamplePanelController(final EVContext context) {
        super(context);
    }

    @Override
    protected String getFxmlFilename() {
        return "/fxml/SamplePanel.fxml";
    }

    @Override
    protected void setupPanel() {

        // setup table columns
        tcTime.setCellValueFactory(new PropertyValueFactory<>("timestamp"));
        tcHeartrate.setCellValueFactory(new PropertyValueFactory<>("heartRate"));
        tcAltitude.setCellValueFactory(new PropertyValueFactory<>("altitude"));
        tcSpeed.setCellValueFactory(new PropertyValueFactory<>("speed"));
        tcDistance.setCellValueFactory(new PropertyValueFactory<>("distance"));
        tcCadence.setCellValueFactory(new PropertyValueFactory<>("cadence"));
        tcTemperature.setCellValueFactory(new PropertyValueFactory<>("temperature"));

        // setup custom number cell factories for all table columns
        final RecordingMode recordingMode = getDocument().getExercise().getRecordingMode();

        tcTime.setCellFactory(new FormattedNumberCellFactory<>(value -> getContext().getFormatUtils()
                .seconds2TimeString(value.intValue() / 1000)));

        tcHeartrate.setCellFactory(new FormattedNumberCellFactory<>(value -> getContext().getFormatUtils()
                .heartRateToString(value.intValue())));

        tcAltitude.setCellFactory(new FormattedNumberCellFactory<>(value -> recordingMode.isAltitude() ? getContext()
                .getFormatUtils().heightToString(value.intValue()) : null));

        tcSpeed.setCellFactory(new FormattedNumberCellFactory<>(value -> recordingMode.isSpeed() ? getContext()
                .getFormatUtils().speedToString(value.floatValue(), 2) : null));

        tcDistance.setCellFactory(new FormattedNumberCellFactory<>(value -> recordingMode.isSpeed() ? getContext()
                .getFormatUtils().distanceToString(value.doubleValue() / 1000d, 3) : null));

        tcCadence.setCellFactory(new FormattedNumberCellFactory<>(value -> recordingMode.isCadence() ? getContext()
                .getFormatUtils().cadenceToString(value.intValue()) : null));

        tcTemperature.setCellFactory(new FormattedNumberCellFactory<>(
                value -> recordingMode.isTemperature() ? getContext().getFormatUtils().temperatureToString(
                        value.shortValue()) : null));

        // set table data
        tvSamples.setPlaceholder(new Label(getContext().getFxResources().getString("pv.info.no_data_available")));
        final ExerciseSample[] samples = getDocument().getExercise().getSampleList();
        tvSamples.setItems(FXCollections.observableArrayList(samples == null ? new ExerciseSample[0] : samples));

        // default sort is the time column
        tvSamples.getSortOrder().add(tcTime);
    }
}

File: st-exerciseviewer/src/main/kotlin/de/saring/exerciseviewer/gui/panels/SamplePanelController.kt
package de.saring.exerciseviewer.gui.panels

import de.saring.exerciseviewer.data.ExerciseSample
import de.saring.exerciseviewer.gui.EVContext
import de.saring.exerciseviewer.gui.EVDocument
import de.saring.util.gui.javafx.FormattedNumberCellFactory
import de.saring.util.unitcalc.TimeUtils
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.control.Label
import javafx.scene.control.TableColumn
import javafx.scene.control.TableView
import javafx.scene.control.cell.PropertyValueFactory

/**
 * Controller (MVC) class of the "Samples" panel, which displays all recorded samples of the exercise in a table.
 *
 * @constructor constructor for dependency injection
 * @param context the ExerciseViewer UI context
 * @param document the ExerciseViewer document / model
 *
 * @author Stefan Saring
 */
class SamplePanelController(
        context: EVContext,
        document: EVDocument) : AbstractPanelController(context, document) {

    @FXML
    private lateinit var tvSamples: TableView<ExerciseSample>

    @FXML
    private lateinit var tcTime: TableColumn<ExerciseSample, Number>
    @FXML
    private lateinit var tcHeartrate: TableColumn<ExerciseSample, Number>
    @FXML
    private lateinit var tcAltitude: TableColumn<ExerciseSample, Number>
    @FXML
    private lateinit var tcSpeed: TableColumn<ExerciseSample, Number>
    @FXML
    private lateinit var tcDistance: TableColumn<ExerciseSample, Number>
    @FXML
    private lateinit var tcCadence: TableColumn<ExerciseSample, Number>
    @FXML
    private lateinit var tcTemperature: TableColumn<ExerciseSample, Number>

    override val fxmlFilename: String = "/fxml/panels/SamplePanel.fxml"

    override fun setupPanel() {

        // setup table columns
        tcTime.cellValueFactory = PropertyValueFactory("timestamp")
        tcHeartrate.cellValueFactory = PropertyValueFactory("heartRate")
        tcAltitude.cellValueFactory = PropertyValueFactory("altitude")
        tcSpeed.cellValueFactory = PropertyValueFactory("speed")
        tcDistance.cellValueFactory = PropertyValueFactory("distance")
        tcCadence.cellValueFactory = PropertyValueFactory("cadence")
        tcTemperature.cellValueFactory = PropertyValueFactory("temperature")

        // setup custom number cell factories for all table columns

        tcTime.cellFactory = FormattedNumberCellFactory {
            if (it != null) TimeUtils.seconds2TimeString(it.toInt() / 1000) else null
        }

        tcHeartrate.cellFactory = FormattedNumberCellFactory {
            if (it != null) context.formatUtils.heartRateToString(it.toInt()) else null
        }

        tcAltitude.cellFactory = FormattedNumberCellFactory {
            if (it != null) context.formatUtils.heightToString(it.toInt()) else null
        }

        tcSpeed.cellFactory = FormattedNumberCellFactory {
            if (it != null) context.formatUtils.speedToString(it.toFloat(), 2, document.speedMode) else null
        }

        tcDistance.cellFactory = FormattedNumberCellFactory {
            if (it != null) context.formatUtils.distanceToString(it.toDouble() / 1000.0, 3) else null
        }

        tcCadence.cellFactory = FormattedNumberCellFactory {
            if (it != null) context.formatUtils.cadenceToString(it.toInt()) else null
        }

        tcTemperature.cellFactory = FormattedNumberCellFactory {
            if (it != null) context.formatUtils.temperatureToString(it.toShort()) else null
        }

        // set table data
        tvSamples.placeholder = Label(context.resources.getString("pv.info.no_data_available"))
        tvSamples.items = FXCollections.observableArrayList(document.exercise.sampleList)

        // default sort is the time column
        tvSamples.sortOrder.add(tcTime)
    }
}
